i18n: Higher-order component for passing translate as prop

aduth opened the pull request:
This pull request seeks to add a new `localize` higher-order component. When called, `localize` returns a new component class with `localize` passed as a prop to the original component when rendered. The advantage of using a higher-order component instead of calling `translate` directly from the `lib/i18n/mixins` module is that the latter does not properly account for `change` events which may be emitted by the `i18nState` object.

__Testing instructions:__

1. Navigate to the [Calypso post editor](http://calypso.localhost:3000/post)
2. Select a site, if prompted
3. Note that the label at the top center of the screen displays "New Post" translated to your current language

blowery commented on client/lib/mixins/i18n/localize.jsx:
```diff
@@ -0,0 +1,35 @@
+		}
+
+		render() {
+			return <ComposedComponent { ...this.props } translate={ translate } />;
```
should we also expose moment / numberFormat?

aduth replied:
>should we also expose moment / numberFormat?

Probably. Will plan to revise.

aduth replied:
>should we also expose moment / numberFormat?

Added in aa04cbc.

blowery commented:
Code LGTM. Can you add a section to the docs on how to use it and why it's a good idea?

aduth commented:
Rebased, added README and tests. @blowery : Is this what you had in mind, or were you thinking to include a note in our top-level `docs/` directory? Doesn't seem like we have much on i18n at all there currently.

blowery commented:
:ship:
